File: 8kyu/(8 kyu) Powers of 2/(8 kyu) Powers of 2.cpp
#include <vector>
#include <cstdint>

// 1 Simple solution
std::vector<uint64_t> powers_of_two(int n) {
    std::vector<uint64_t> result;
    result.reserve(n + 1);
    for (int i = 0; i <= n; ++i) {
        result.push_back(static_cast<uint64_t>(1) << i);
    }
    return result;
}

// 2 Optimized solution
std::vector<uint64_t> powers_of_two(int n) {
    std::vector<uint64_t> result(n + 1);
    for (int i = 0; i <= n; ++i) {
        result[i] = 1ULL << i;
    }
    return result;
}

// 3 Clever solution
#include <algorithm>
#include <iterator>
#include <numeric>
std::vector<uint64_t> powers_of_two(int n) {
    std::vector<uint64_t> result(n + 1);
    std::iota(result.begin(), result.end(), 0);
    std::transform(result.begin(), result.end(), result.begin(),
                   [](int i) { return 1ULL << i; });
    return result;
}

File: 8kyu/(8 kyu) Powers of 2/(8 kyu) Powers of 2.go
package kata

// 1 Simple solution
func PowersOfTwo(n int) []uint64 {
	result := make([]uint64, n+1)
	var power uint64 = 1
	for i := 0; i <= n; i++ {
		result[i] = power
		power *= 2
	}
	return result
}

// 2 Optimized solution
func PowersOfTwo(n int) []uint64 {
	result := make([]uint64, n+1)
	for i := 0; i <= n; i++ {
		result[i] = 1 << i
	}
	return result
}

// 3 Clever solution
func PowersOfTwo(n int) (arr []uint64) {
	for i := 0; i <= n; i++ {
		arr = append(arr, 1 << i)
	}
	return
}